apps/OboeTester/app/src/main/cpp/jni-bridge.cpp

```diff
@@ -1126,54 +1126,159 @@ Java_com_mobileer_oboetester_AudioWorkloadTestActivity_close(JNIEnv *env, jobjec
+        return;
     }
 
-    jmethodID arrayListAdd = env->GetMethodID(arrayListClass, "add", "(Ljava/lang/Object;)Z");
-    if (arrayListAdd == nullptr) {
-        std::cerr << "Error: Could not find 'add' method for ArrayList" << std::endl;
+    // Delete global references
+    if (g_callbackStatusClass != nullptr) {
+        env->DeleteGlobalRef(g_callbackStatusClass);
+        g_callbackStatusClass = nullptr;
+    }
+    if (g_arrayListClass != nullptr) {
+        env->DeleteGlobalRef(g_arrayListClass);
+        g_arrayListClass = nullptr;
+    }
+
+    g_callbackStatusConstructor = nullptr;
+    g_arrayListConstructor = nullptr;
+    g_arrayListAddMethod = nullptr;
+
+    g_javaVM = nullptr;
+    std::cout << "JNI_OnUnload: Released global JNI references." << std::endl;
+}
+
+JNIEXPORT jobject JNICALL
+Java_com_mobileer_oboetester_AudioWorkloadTestActivity_getCallbackStatistics(JNIEnv *env,
+                                                                             jobject obj) {
+    if (g_callbackStatusClass == nullptr || g_callbackStatusConstructor == nullptr ||
+        g_arrayListClass == nullptr || g_arrayListConstructor == nullptr ||
+        g_arrayListAddMethod == nullptr) {
+        std::cerr << "Error: JNI IDs not cached. Initialization in JNI_OnLoad might have failed."
+                << std::endl;
         return nullptr;
     }
 
-    jobject javaList = env->NewObject(arrayListClass, arrayListConstructor);
+    std::vector<AudioWorkloadTest::CallbackStatus> cppCallbackStats =
+            sAudioWorkload.getCallbackStatistics();
+
+    jobject javaList = env->NewObject(g_arrayListClass, g_arrayListConstructor);
+    if (javaList == nullptr) {
+        std::cerr << "Error: Could not create new ArrayList object." << std::endl;
+        if (env->ExceptionCheck()) env->ExceptionDescribe();
+        return nullptr;
+    }
 
     for (const auto& status : cppCallbackStats) {
-        // Must match CallbackStatus constructor in AudioWorkloadTestActivity.java
         jobject javaStatus = env->NewObject(
-                callbackStatusClass,
-                callbackStatusConstructor,
+                g_callbackStatusClass,
+                g_callbackStatusConstructor,
                 (jint)status.numVoices,
                 (jlong)status.beginTimeNs,
                 (jlong)status.finishTimeNs,
                 (jint)status.xRunCount,
                 (jint)status.cpuIndex
         );
-        env->CallBooleanMethod(javaList, arrayListAdd, javaStatus);
-        env->DeleteLocalRef(javaStatus); // Important: Release local references
+        if (javaStatus == nullptr) {
+            std::cerr << "Error: Could not create new CallbackStatus object." << std::endl;
+            if (env->ExceptionCheck()) env->ExceptionDescribe();
+            env->DeleteLocalRef(javaList);
+            return nullptr;
+        }
+
+        env->CallBooleanMethod(javaList, g_arrayListAddMethod, javaStatus);
+        if (env->ExceptionCheck()) {
```

Comment:
Given it is an arraylist add, I'm not sure what exception may happen here. Could you explain more details about your concerns?

Reply:
We use it quite liberally within our code in Android. I don't see any issues with extra exception checks as we could have easily typed the signature wrong for g_arrayListAddMethod. This makes it so when others copy paste this, they can make sure the type is right

Reply:
It is not wrong to check the exception... But just if there is not exception to be thrown, why we bother to check exception...

If you type wrong signature of `g_arrayListAddMethod`, it should be caught when finding the method instead of here...

Reply:
See https://www.w3schools.com/java/ref_arraylist_add.asp

ArrayList.add can still return IndexOutOfBoundsException although it should never return it in this case